### PR TITLE
KAFKA-14084: SCRAM support in KRaft.

### PR DESCRIPTION
**Pull request by pprovenzano**
This commit adds support to store the SCRAM credentials in a cluster with KRaft quorum servers and no ZK cluster backing the metadata.

This commit does not include changes to allow bootstrap of a KRaft cluster with brokers using SCRAM at bootstrap time. The bootstrap support will come with a separate KIP and commit.

Tests pass along with manual testing of credential operations against a cluster with a separate advertised listener for SASL_SCRAM

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cmccabe** [comment]
Can we have a test in `ControllerApisTest`? These are mainly to verify that our authorization logic is sound.

**Comment by cmccabe** [comment]
We should not be changing `BrokerMetadataSnapshotterTest.scala` here -- doesn't seem needed

**Comment by cmccabe** [comment]
* We need a `ScramControlManagerTest` as well

**Comment by pprovenzano** [comment]
> 

The changes to `BrokerMetadataSnapshotterTest.scala` are needed and small. The size of the SCRAM records is larger than 1024 bytes and so I increased it to 4096 bytes in the max allocation size in the mock snapshotter instantiated in the test.  I also added a timeout to these tests because the failure mode of the records being to large is for the test to hang indefinitely. I had a very hard time figuring out which test was hanging without it.

**Comment by pprovenzano** [comment]
I just updated AlterUserScramCredentialsRequestTest to work in KRaft after finding it while looking into the tests you wanted. It covers a lot of what you want testing ScramControlManager and ControllerApis given that it is an end to end test.

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala]: unwanted line

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: maybe we can use `import org.apache.kafka.image._`

**Comment by omkreddy** [comment]
As per this comment https://github.com/apache/kafka/pull/12196#issuecomment-1148690209, Do we need to handle `handleDescribeUserScramCredentialsRequest ` 

Also are we going to update existing integration tests after bootstrap support?

**Comment by pprovenzano** [comment]
I plan to update existing tests after bootstrap support is committed.
I'll check on the `handleDescribeUserScramCredentialsRequest`

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (omkreddy) [inline core/src/main/scala/kafka/server/ControllerApis.scala:823]: We need to authorisation `authHelper.authorizeClusterOperation(request, ALTER)` check here.

Similar check in KafkaAPI:
[handleAlterUserScramCredentialsRequest](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L3326)

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala
Comment (omkreddy) [inline core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:229]: May I know, how are applying changes on Controller Nodes: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ControllerServer.scala#L176

File: metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java:112]: I think, we should not return sensitive info like salt, saltedPassword etc..

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramImage.java:169]: we should not print  sensitive info like `userMap.get(`

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramImage.java:53]: ImageWriterOptions is unused.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala
Comment (omkreddy) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:229]: I think we need to update `ScramControlManager.reply` method to call `credentialProvider. updateCredential(), removeCredentials()` methods

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: `scramControl` => `scramControlManager`

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala:49]: we should also support Describe API and update `testAlterAndDescribe` method

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:144]: We should return sensitive data

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:314]: unused method

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:167]: This method is  important for applying changes. Can we add few unit tests scenarios to verify this method.

**Review by omkreddy — COMMENTED** [review]
@pprovenzano Thanks for the PR. left few comments

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/ControllerApis.scala:823]: Lets chat about this. Currently talking directly with the controller is not allowed for these operations. Only brokers are allowed.

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:167]: AlterUserScramCredentialsRequestTest.scala does that for the error cases.

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:53]: This is the same as in `AclsImage.java`

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:229]: The credentials are added to the metadata log and then consumed by the brokers and added to the broker cache.

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:144]: As far as I can tell, the toString method is never used. 
I think it is useful for debugging, but I'd like to know how we handle this in general.

File: metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java:112]: As far as I can tell, the toString method is never used.
I think it is useful for debugging, but I'd like to know how we handle this in general.

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:169]: As far as I can tell, the toString method is never used.
I think it is useful for debugging, but I'd like to know how we handle this in general.

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (pprovenzano) [reply core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala]: Fix

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (pprovenzano) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Fixed

File: metadata/src/main/java/org/apache/kafka/controller/QuorumController.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/controller/QuorumController.java]: Fixed

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:314]: Keeping this as it may be useful for the bootstrap code.

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java]: "(salt= [hidden], saltedPassword= [hidden]", iterations=[hidden]";

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:169]: This one should b okay. It is just printing a list of usernames and which SCRAM mechanism they use. No passwords or salts displayed here.

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/ControllerApis.scala:823]: I understand. Thanks for your explanation. This is now fixed.

File: metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/controller/ScramControlManager.java:144]: Fixed

File: metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramCredentialData.java:112]: Fixed

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:169]: Fixed

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (pprovenzano) [reply core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala:49]: I've added Describe and updated the test.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaApis.scala:3631]: maybe we can keep them for now

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (omkreddy) [inline core/src/main/scala/kafka/server/ControllerApis.scala:102]: I think, Similar to other APIs (CREATE_ACLS, DESCRIBE_ACLS, etc), we should also support DescribeUserScramCredentialsRequest API here.

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala]: we generally use 'TestUtils.waitForCondition` in tests to wait for condition

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramImage.java]: Can we use same error message as https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ZkAdminManager.scala#L802

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramImage.java:113]: Dont we need to check for ScramMechanism.UNKNOWN
similar to https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ZkAdminManager.scala#L812 and 
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ZkAdminManager.scala#L816

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (omkreddy) [inline metadata/src/main/java/org/apache/kafka/image/ScramImage.java:124]: Do we have separate error message for "Username must not be empty"

**Review by omkreddy — COMMENTED** [review]
We also need to update [DescribeUserScramCredentialsRequestTest.scala]

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:113]: No, in our implementation we will never create a `Map<String, ScramCredentialData>` for ScramMechanism.UNKNOWN because we never populate any credentials for an unknown mechanism. Thus I can iterate over the maps knowing that they are all valid.

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:124]: The request contains a list of users. If the list is empty then the request is to describe all the users. I personally think this is a security issue but I'm just implementing what was there for Zk. 

If there are no users with SCRAM credentials and the request is to describe all the users, then an empty response is returned. It is not an error. This is tested in the unit test.

Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:124]: I'm not sure why there is an illegalUsers error in the Zk case as it explicitly tests for users.get.isEmtpy for the describe all case. https://github.com/apache/kafka/blob/b9754747d6eaa029c4bb69b073d749ff8df15908/core/src/main/scala/kafka/server/ZkAdminManager.scala#L805

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java]: Yes!

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3631]: The compiler throws an error if we try to keep them.

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/ControllerApis.scala:102]: We cannot test it until we update ControllerApis to handle METADATA requests, so I'm going to wait on adding it.

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (pprovenzano) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java]: Fixed!

File: core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala]: can we remove unused import?

File: core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala
Comment (omkreddy) [reply core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala]: Can we address this?

Comment (pprovenzano) [reply core/src/test/scala/unit/kafka/server/AlterUserScramCredentialsRequestTest.scala]: Fixed

File: core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala
Comment (pprovenzano) [reply core/src/test/scala/unit/kafka/server/DescribeUserScramCredentialsRequestTest.scala]: Fixed

**Review by omkreddy — APPROVED** [review]
@pprovenzano Thanks for the Review, LGTM. I will merge after getting review from KRaft experts

Some of the pending items:
1. Updating CredentialProvider in Controller
2. Support for DescribeUserScramCredentialsRequest in Controller
3. KIP-900 Implementation
4. Implement Integration tests Kraft Scram
5. Update System tests

File: core/src/main/scala/kafka/server/ControllerApis.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/server/ControllerApis.scala:102]: Yeah. Just to restate what I think we all know here, clients ask brokers to describe SCRAM credentials, rather than controllers. So this doesn't technically need to be implemented, except for testing.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala
Comment (cmccabe) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:229]: I think we should split out this code for applying the SCRAM stuff into a separate `MetadataPublisher`, as we did with `DynamicConfigPublisher`. Then the controller can run it as well.

But we can do that in a follow-on change.

File: metadata/src/main/java/org/apache/kafka/image/ScramImage.java
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/ScramImage.java:53]: You need to check the MetadataVersion and invoke `options.handleLoss` if it is too low. See my comment below. This feature will only be supported in `IBP_3_5_IV0` and newer.

Also we'll need a test for this (that we throw an exception when required if SCRAM is present). That should be in `./metadata/src/test/java/org/apache/kafka/image/ScramImageTesst.java` (which still needs to be written) :)

**Comment by cmccabe** [comment]
We need a test like `./metadata/src/test/java/org/apache/kafka/image/ScramImageTesst.java` , similar to the tests for the other Image classes.

**Comment by cmccabe** [comment]
Thanks for this, @pprovenzano . It all looks good, but there is one change that we really do need here: we need to add SCRAM to MetadataVersion. I suspect we can add it to `IBP_3_5_IV0`. This also means having `MetadataVersion#isScramSupported`, etc.

**Comment by pprovenzano** [comment]
> We need a test like `./metadata/src/test/java/org/apache/kafka/image/ScramImageTesst.java` , similar to the tests for the other Image classes.

I'm confused by this comment. There is a `./metadata/src/test/java/org/apache/kafka/image/ScramImageTest.java` in the review. It is the last file in the review.

**Comment by cmccabe** [comment]
> I'm confused by this comment. There is a ./metadata/src/test/java/org/apache/kafka/image/ScramImageTest.java in the review. It is the last file in the review.

Sorry, my mistake. I see it now. I don't know how I missed it.

We still need the MetadataVersion fixes and a test for that in the image test.

File: core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala
Comment (pprovenzano) [reply core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:229]: https://issues.apache.org/jira/browse/KAFKA-14775

**Comment by pprovenzano** [comment]
I've added the MetadataVersion fixes with tests.

File: metadata/src/test/java/org/apache/kafka/image/ScramImageTest.java
Comment (cmccabe) [inline metadata/src/test/java/org/apache/kafka/image/ScramImageTest.java:156]: You should be able to catch this with assertThrows and inspect the exception a bit. But we can do this in a follow-on.

**Review by cmccabe — APPROVED** [review]
LGTM. Thanks @pprovenzano

**Comment by cmccabe** [comment]
follow-ons: 
https://issues.apache.org/jira/browse/KAFKA-14765
https://issues.apache.org/jira/browse/KAFKA-14775
https://issues.apache.org/jira/browse/KAFKA-14776

**Pull request merged**
